Prevent scalers to scale near-constant features very large values

ogrisel opened the pull request:
Fixes #19450 (an edge case of a test initially written for #19426).

The original problem happens when fitting `StandardScaler(with_mean=False)`  with `sample_weight` on constant (but non-zero) columns. Because of the sample weight, the variance is not exactly zero but a very small value.

This PRs tries to solve the problem by using a dtype-dependent eps threshold to detect constant features instead of a hard `var == 0.` test. I added more tests and also adapted `sklearn.linear_model._base._preprocess_data` so that linear models with `normalize=True` behave consistently.

Remaining issues:

- `sklearn.linear_model._base._preprocess_data` is equivalent to StandardScaler up to a factor of `np.sqrt(X.shape[0])` for all features except the ones that constant and non-zero. Those features are typically useless and get absorbed into the intercept but because of the interaction with regularization, the matching coef and intercept value will be different, even if the predictions are the same. Maybe this is not a problem. /cc @agramfort @maikia.

- There a numerical stability issue to compute the variance of large but constant features on sparse data: see the tests marked `XFAIL`, for instance with a column with a constant value of `100.`, we get a variance that is far from zero (depends on `n_samples`). This should probably be tackled in a separate PR and tracked by a dedicated bug report in #19546.

- Last constant features would stay large constant features with `StandardScaler`. This might feel surprising for some users. Shall we warn when we detect those? This would risk spurious warnings when cross-validating on small-ish datasets which could be annoying in practice. **Update:** we discussed this point at the last dev meeting and the opinion of letting those feature passthough was shared. Maybe we should add a waning but this can be tackled in a dedicated PR. I opened an issue to track the discussion on this point: #19547.

ogrisel committed:
Fix scaler on near-constant features

ogrisel committed:
Remove useless import

ogrisel committed:
Update changelog

thomasjpfan commented:
I am guessing we can not use `ptp`, like we do in `VarianceThreshold`, because of the sample weights?

https://github.com/scikit-learn/scikit-learn/blob/1000d0a61be311542e01d56f6745178307406395/sklearn/feature_selection/_variance_threshold.py#L72-L86

ogrisel commented:
> I am guessing we can not use ptp, like we do in VarianceThreshold, because of the sample weights?

That's interesting, but indeed I am not sure what to do with sample weights and the problem is most often triggered when using sample weights in practice.

maikia commented on sklearn/preprocessing/_data.py:
```diff
@@ -60,22 +60,27 @@
 
 
-def _handle_zeros_in_scale(scale, copy=True):
+def _handle_zeros_in_scale(scale, copy=True, constant_mask=None):
```
maybe add a short doc on what is a parameter `constant_mask`

ogrisel replied:
Done in 139d37d

maikia reviewed (commented):
LGTM thanks for taking care of this issue @ogrisel

ogrisel commented on sklearn/preprocessing/_data.py:
```diff
@@ -60,22 +60,27 @@
 
 
-def _handle_zeros_in_scale(scale, copy=True):
+def _handle_zeros_in_scale(scale, copy=True, constant_mask=None):
```
```suggestion
def _handle_zeros_in_scale(scale, copy=True, constant_mask=None):
    """Set scales of near constant features to 1.

    The goal is to avoid division by very small or zero values.

    Near constant features are detected automatically by identifying
    scales close to machine precision unless they are precomputed by
    the caller. Typically for standard scaling, the scales are the standard
    deviation while near constant features are better detected on the
    computed variances which are closer to machine precision by
    construction.
    """
```

ogrisel commented on sklearn/preprocessing/_data.py:
```diff
@@ -60,22 +60,27 @@
 
-def _handle_zeros_in_scale(scale, copy=True):
+def _handle_zeros_in_scale(scale, copy=True, constant_mask=None):
     """Makes sure that whenever scale is zero, we handle it correctly.
 
     This happens in most scalers when we have constant features.
     """
```
```suggestion
```suggestion
def _handle_zeros_in_scale(scale, copy=True, constant_mask=None):
    """Set scales of near constant features to 1.

    The goal is to avoid division by very small or zero values.

    Near constant features are detected automatically by identifying
    scales close to machine precision unless they are precomputed by
    the caller. Typically for standard scaling, the scales are the standard
    deviation while near constant features are better detected on the
    computed variances which are closer to machine precision by
    construction.
    """
```
```

ogrisel commented on sklearn/linear_model/tests/test_base.py:
```diff
@@ -478,10 +478,8 @@ def test_preprocess_data_weighted(is_sparse):
-    # https://github.com/scikit-learn/scikit-learn/issues/19450
-    # X[:, 2] = 1.
+    # Constant non-zero feature.
+    X[:, 2] = 1.
```
Thanks for the feedback. This is indeed a problem. I will open a dedicated issue, both for standard scaler and this which both are using the same private API.

ogrisel committed:
Update sklearn/preprocessing/_data.py

ogrisel commented on sklearn/preprocessing/_data.py:
```diff
@@ -60,22 +60,36 @@
-
-    This happens in most scalers when we have constant features.
+def _handle_zeros_in_scale(scale, copy=True, constant_mask=None):
+    """Set scales of near constant features to 1.
+
+    The goal is to avoid division by very small or zero values.
+
+    Near constant features are detected automatically by identifying
+    scales close to machine precision unless they are precomputed by
+    the caller and passed with the `constant_mask` kwarg.
+    
+    Typically for standard scaling, the scales are the standard
+    deviation while near constant features are better detected on the
+    computed variances which are closer to machine precision by
+    construction.
     """
```
```suggestion
    """Set scales of near constant features to 1.

    The goal is to avoid division by very small or zero values.

    Near constant features are detected automatically by identifying
    scales close to machine precision unless they are precomputed by
    the caller and passed with the `constant_mask` kwarg.

    Typically for standard scaling, the scales are the standard
    deviation while near constant features are better detected on the
    computed variances which are closer to machine precision by
    construction.
    """
```

ogrisel committed:
Update sklearn/preprocessing/_data.py

ogrisel commented:
@thomasjpfan @maikia @agramfort I addressed the comments that I think are directly related to the PR and opened dedicated issues for follow-up work: #19546 and #19547.

I think this PR could be merged as it is without waiting for either of the follow-up issues as it's already a bug fix as it stands.

agramfort reviewed (approved):
LGTM

thx @ogrisel and @maikia

ogrisel commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -414,6 +414,61 @@ def test_standard_scaler_dtype(add_sample_weight, sparse_constructor):
+@pytest.mark.parametrize("constant", [0, 1., 100.])
+def test_standard_scaler_constant_features(
+        scaler, add_sample_weight, sparse_constructor, dtype, constant):
+    if (isinstance(scaler, StandardScaler)
+            and constant > 1
+            and sparse_constructor is not None
+            and add_sample_weight):
+        pytest.xfail("Computation of weighted variance is numerically unstable"
+                     " for sparse data")
```
```suggestion
    if (isinstance(scaler, StandardScaler)
            and constant > 1
            and sparse_constructor is not None
            and add_sample_weight):
        # https://github.com/scikit-learn/scikit-learn/issues/19546
        pytest.xfail("Computation of weighted variance is numerically unstable"
                     " for sparse data. See: #19546.")
```

ogrisel committed:
Update sklearn/preprocessing/tests/test_data.py

maikia reviewed (approved):
LGTM too. Thanks @ogrisel

rth reviewed (approved):
Thanks for fixing this!
